features/configuration/read_options_from_file.feature

```diff
@@ -9,12 +9,18 @@ Feature: read command line configuration options from files
     * Project:  `./.rspec` (i.e. in the project's root directory, usually
       checked into the project)
 
-    * Global: `~/.rspec` (i.e. in the user's home directory)
+    * Global (HOME): `~/.rspec` (i.e. in the user's home directory)
 
-  Configuration options are loaded from `~/.rspec`, `.rspec`, `.rspec-local`,
-  command line switches, and the `SPEC_OPTS` environment variable (listed in
-  lowest to highest precedence; for example, an option in `~/.rspec` can be
-  overridden by an option in `.rspec-local`).
+    * Global (XDG): `$XDG_CONFIG_HOME/rspec/options` (i.e. in the user's
+      [the XDG Base Directory
+      Specification](https://specifications.freedesktop.org/basedir-spec/latest/)
+      config directory)
+
+  Configuration options are loaded from `$XDG_CONFIG_HOME/rspec/options`,
+  `~/.rspec`, `.rspec`, `.rspec-local`, command line switches, and the
+  `SPEC_OPTS` environment variable (listed in lowest to highest precedence; for
+  example, an option in `~/.rspec` can be overridden by an option in
+  `.rspec-local`).
```

Comment:
Generally I believe they are cumulative (e.g. a global `.rspec` and local `.rspec` combine), indeed I would expect a local `.rspec` file to override this.

Reply:
Right, the other existing file locations are cumulative. But I think XDG location should be considered the _replacement_ for `~/.rspec`. (And I presume ~/.rspec will eventually be deprecated in favor of the xdg-defined location.) As such, I feel like making these two cumulative sends the wrong message.

Reply:
It will not be the replacement for `~/.rspec`,  not all systems utilise it. For example, OS X, Windows, therefore this can only be considered additionally supported.

Reply:
I think what @jasonkarns is saying makes sense.  The 3 existing config files have specific purposes for all existing:

* `~/.rspec` is for personal options that apply to all RSpec projects
* `.rspec` is for team options that apply to a specific project
* `./rspec-local` is for personal options that apply to a specific project

My understanding of XDG (and this PR) is that it's a more standardized system for providing what we're using `~/.rspec` for.  As such, for a given RSpec run, it doesn't really make sense (IMO) to merge `~/.rspec` and the XDG rspec config file.  If the user is setup to use the XDG config file, I think we should use it _instead_ of the `~/.rspec` file.

That said, I don't think the presence of the `XDG_CONFIG_HOME` env var is a sufficient signal to use the XDG file and ignore `~/.rspec`, because someone could have `XDG_CONFIG_HOME` set for use by another tool, without realizing that RSpec will be using it, and could also have a `~/.rspec` file that they expect to be honored.  If we ignored `~/.rspec` because `XDG_CONFIG_HOME` happened to be set on their machine it'd be bad.

So maybe the solution is to use the XDG config file if a file can be found, using either the env var or the default XDG path.  If it can't be found, we fall back to `~/.rspec`.

> For example, OS X, Windows, therefore this can only be considered additionally supported.

My understanding (which may be wrong; I haven't read much about XDG) is that it's not an OS-specific thing.  It's a standard that any tool can choose to honor, and any user can choose to use on any OS.  If a user sets `XDG_CONFIG_HOME` and drops a config file in the configured path, and the tool is documented as supporting XDG config, it should load and use the user's config file, regardless of what OS they are on.

Reply:
You can retrofit it into OS X and to some extent Windows (Windows is quite hostile to environment variables), but its not an existing OS established pattern like it is on Linux, thus I am against *deprecating* `~/.rspec` in favour of it. I'm quite happy for us to check for it and use it without cumulatively combining it with a `~/.rspec` file, but we should issue a warning when both files exist.

Reply:
Agreed, issuing a warning when we ignore `~/.rspec` would be a good idea.  And yeah, I don't know enough about XDG to know if deprecating `~/.rspec` would make sense or not.  My instinct is not to consider deprecating until we have some usage data and user feedback.

Reply:
> So maybe the solution is to use the XDG config file if a file can be found, using either the env var or the default XDG path. If it can't be found, we fall back to ~/.rspec.

Strong agree. A lot of tools have partial (non-compliant) XDG implementations where they only trigger XDG lookup if the XDG var is set. Which means one can't just use the default locations specified in the spec without setting the variables. Therefore it's quite likely for some users to have XDG_* vars set _without_ expecting them to be used (yet) for rspec.

> My understanding (which may be wrong; I haven't read much about XDG) is that it's not an OS-specific thing.

Also agree. Multiple cross-platform tools leverage it safely (like git). Indeed, the XDG defaults all live in `$HOME`, so even without setting the XDG vars, users can opt-in by just moving their rspec configs into whatever their OS considers `$HOME` (+ `/rspec/config`). And moving off the defaults is still just a matter of setting the env var as desired.

(I personally am on OSX have a dozen or more tools adhering to XDG.)

Reply:
👍 I just didn't want our assumed default here to be that we would deprecate the existing setup in favour of this :)

Reply:
Works for me. Especially since rspec doesn't do any generation of the system-wide rspec file. (to my knowledge?)

My main concern for it to be deprecated was reminiscent of other tools that _do_ generate their config files, and continued to write directly to `~/.foo` despite `XDG_CONFIG_HOME` being set.

As long as rspec respects `${XDG_CONFIG_HOME:-$HOME/.config}/rspec/options` when it exists, I'm :+1: 😀

Reply:
Isn't it's easier to explain it as yet another file in the chain rather than a replacement for the HOME file if present?

What advantage do we get from ignoring the HOME file if both files exist?

I would have both files until all my projects were on a version of rspec that supports XDG, but they would be identical (HOME file would be a symlink to the XDG file) so I don't think I'd get anything from that.

If there are some tools out there that generates a HOME file, then it might be confusing for a user to get an effect not present in their XDG file, but it could also be confusing if they didn't get the effect despite the tool claiming to have installed the options. I cannot tell which is worse.

Reply:
I'm not sure that the explanation is sufficiently more confusing to warrant making the files semantically ambiguous. The XDG location is semantically a replacement for the root home directory. Going against that would be a violation of the spirit of the XDG spec, IMO, if not a true violation of the spec. And I agree with @myronmarston 's comment that there are three distinct roles for the different files.

> The 3 existing config files have specific purposes for all existing:
> 
> ~/.rspec is for personal options that apply to all RSpec projects
> .rspec is for team options that apply to a specific project
> ./rspec-local is for personal options that apply to a specific project

The XDG location is precisely the same role as ~/.rspec: "personal options that apply to all RSpec projects". Having two different files fill the same role would be more confusing, IMO.

Reply:
> Having two different files fill the same role would be more confusing, IMO.

Agree.  In addition, the precedence between the 3 existing files is easy to understand and make sense (e.g. personal options can override team ones, and local personal options can override global personal options).  But with this new file fulfilling the same role as `~/.rspec`, there's no basis for having `~/.rspec` override the XDG file, or vice versa.  It's arbitrary.  I'd prefer not to make an arbitrary decision.